### ISSUE TITLE
TiDB query return wrong result when using server prepare statement

### ISSUE THREAD

**Issue by VoxT** [pingcap/tidb#35040]
### 1. Minimal reproduce step (Required)
- Create table
```sql
CREATE TABLE `services` (
  `id` bigint(20) NOT NULL,
  `name` varchar(20) NOT NULL,
  `password` varchar(100) NOT NULL,
  `created_at` bigint(20) DEFAULT NULL,
  `updated_at` bigint(20) DEFAULT NULL,
  `env` varchar(10) NOT NULL,
  `is_enabled` tinyint(1) DEFAULT '1',
  `is_realtime` tinyint(1) NOT NULL DEFAULT '1',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `password_unique` (`password`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin

INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (123, 'Hello', 'Hello', 20, 20, 'Hello', 1, 1);
INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (20001, 'FABO', 'FA', 1574306994155, 1574306994155, 'DEV', 1, 0);
INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (20002, 'UM', 'UM', 1574306994155, 1574306994155, 'DEV', 1, 1);
INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (20003, 'ZM', 'ZA', 1574306994155, 1574306994155, 'DEV', 1, 1);
INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (20004, 'BK', 'BK', 1574306994155, 1574306994155, 'DEV', 1, 1);
INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (20006, 'QC', 'QC', 1574306994155, 1574306994155, 'DEV', 1, 1);
INSERT INTO services(id, name, password, created_at, updated_at, env, is_enabled, is_realtime) VALUES (20009, 'INTERNAL', 'INTERNAL', 1574306994155, 1574306994155, 'INTERNAL', 1, 0);

```

- Prepare the following statement
```sql
prepare stmt1 FROM 'SELECT * FROM services where 1=1 and id = ?';

```

- Query id=20001
```sql
SET @a = 20001;
execute stmt1 using @a;
 => Result id= 20001  (CORRECT RESULT)
```
- Query id=20002
```sql
SET @a = 20002;
execute stmt1 using @a;
 => Result: id=`20001`  (WRONG RESULT)
```

### 2. What did you expect to see? (Required)
- Query id=20002=> Result: id=`20002` 

### 3. What did you see instead (Required)
- Query id=20002 => Result: id=`20001` 

I think there's a bug for TiDB execution plan when the query has ` 1=1 ` condition , when i remove that condition, the query work correctly.

### 4. What is your TiDB version? (Required)
```code
Release Version: v5.4.1
Edition: Community
Git Commit Hash: be484475e76b3aaa2cad520111353c3327db35fa
Git Branch: heads/refs/tags/v5.4.1
UTC Build Time: 2022-05-07 04:31:32
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

**Comment by VoxT** [comment]
@tiancaiamao could you take a look on this.

**Comment by tiancaiamao** [comment]
I test it on master and find no reproduce.
I believe this has been fixed by https://github.com/pingcap/tidb/pull/34739/

**Comment by tiancaiamao** [comment]
Duplicated with https://github.com/pingcap/tidb/issues/33628

**Comment by tiancaiamao** [comment]
prepared plan cache is still an expiremental feature in 5.4.X, so I'm not sure whether this would be backported to 5.4 or even 6.0
You can keep an eye on the cherry-pick PR https://github.com/pingcap/tidb/pull/34782 to see  whether it'll be fix on 5.4.X, or use the coming 6.1 version.
6.1 will be a LTS version and prepared plan cache is planned to be enable by default.  @VoxT

**Comment by VoxT** [comment]
Thank you @ti-chi-bot.
By the way, is there any way to avoid this in production? some workaround?

**Comment by tiancaiamao** [comment]
> Thank you @ti-chi-bot. By the way, is there any way to avoid this in production? some workaround?

Workaround: don't use it...
Set here to `false` https://github.com/pingcap/tidb/blob/4d888ce22b08b9d2bbc2b66890d35195f666e90b/config/config.toml.example#L322

**Comment by VoxT** [comment]
@tiancaiamao 
We already set this config to false, but this still happens?

**Comment by VoxT** [comment]
> > @tiancaiamao We already set this config to false, but this still happens?
> 
> That sounds unreasonable, you can use this to verify the config:
> 
> ```
> curl http://127.0.0.1:10080/config |grep prepared-plan-cache
> ```

Here the result, it's already disabled:
```
"prepared-plan-cache":{"enabled":false,"capacity":1000,"memory-guard-ratio":0.1}
```

**Comment by tiancaiamao** [comment]
@qw4990 Do you have any idea how this could happen? prepared-plan-cache is disabled.

**Comment by qw4990** [comment]
@VoxT Sorry for my late reply. I guess it's caused by these two issues:
1. https://github.com/pingcap/tidb/pull/34579: the optimizer cannot validate some point-get plans correctly when reusing them in some cases;
2. https://github.com/pingcap/tidb/pull/34739: the prep-plan-cache switch cannot control point-get plans;

Since the prep-plan-cache switch cannot control point-get plans, so we cannot avoid this problem by disabling plan-cache directly.
I reproduced this issue locally and found an ugly workaround 🤦 : modifying the prep-statement to `prepare stmt1 FROM 'SELECT * FROM services where 1=1 and id>?-1 and id<?+1` to make the optimizer not generate a point-get plan for this query.

As @tiancaiamao said, I think this is a better solution:
`You can keep an eye on the cherry-pick PR https://github.com/pingcap/tidb/pull/34782 to see whether it'll be fix on 5.4.X, or use the coming 6.1 version.
6.1 will be a LTS version and prepared plan cache is planned to be enable by default`

**Comment by VoxT** [comment]
Thank you @qw4990, @tiancaiamao I will  keep an eye on that
